scripts/http-security-headers.nse

```diff
@@ -0,0 +1,305 @@
+-- <elem key="Description">Will prevent the browser from MIME-sniffing a response away from the declared content-type.</elem>
+-- <elem>Content-Security-Policy is configured.</elem>
+-- <elem key="Header">Content-Security-Policy: script-src 'self'</elem>
+-- <elem key="Description">Loading policy for all resources type in case of a resource type dedicated directive is not defined (fallback).</elem>
+-- <elem>X-Permitted-Cross-Domain-Policies are configured.</elem>
+-- <elem key="Header">X-Permitted-Cross-Domain-Policies: none</elem>
+-- <elem key="Description">No policy files are allowed anywhere on the target server, including this master policy file.</elem>
+--
+-- @args http-security-headers.path The URL path to request. The default path is "/".
+
+author = {"Icaro Torres", "Vinamra Bhatia"}
+license = "Same as Nmap--See https://nmap.org/book/man-legal.html"
+categories = {"discovery", "safe"}
+
+portrule = shortport.port_or_service({80,443}, "http", "tcp")
+
+local function fail (err) return stdnse.format_output(false, err) end
+
+action = function(host, port)
+  local path = stdnse.get_script_args(SCRIPT_NAME..".path") or "/"
+  local response
+  local output_info = {}
+  local hsts_header = {}
+  local hpkp_header = {}
+  local xframe_header = {}
+  local x_xss_header = {}
+  local x_content_type_header = {}
+  local csp_header = {}
+  local x_cross_domain_header = {}
+  local cookie = {}
+  local req_opt = {redirect_ok=function(host,port)
+    local c = 2
+    return function(uri)
+      if ( c==0 ) then return false end
+      c = c - 1
+      return true
+    end
+  end}
+
+  response = http.head(host, port, path, req_opt)
+
+  output_info = stdnse.output_table()
+
+  if response == nil then
+    return fail("Request failed")
+  end
+
+  if response.rawheader == nil then
+    return fail("Response didn't include a proper header")
+  end
+
+  for _, line in pairs(response.rawheader) do
+    if line:match("[Ss]trict.[Tt]ransport.[Ss]ecurity") then
+      table.insert(hsts_header, line)
+    end
+    if line:match("[Pp]ublic.[Kk]ey.[Pp]ins") then
+      table.insert(hpkp_header, line)
+    end
+    if line:match("[Xx].[Ff]rame.[Oo]ptions") then
+      table.insert(xframe_header, line)
+    end
+    if line:match("[Xx].[Xx][Ss][Ss].[Pp]rotection") then
+      table.insert(x_xss_header, line)
+    end
+    if line:match("[Xx].[Cc]ontent.[Tt]ype.[Oo]ptions") then
+      table.insert(x_content_type_header, line)
+    end
+    if line:match("[Cc]ontent.[Ss]ecurity.[Pp]olicy") then
+      table.insert(csp_header, line)
+    end
+    if line:match("[Xx].[Pp]ermitted.[Cc]ross.[Dd]omain.[Pp]olicies") then
+      table.insert(x_cross_domain_header, line)
+    end
+    if line:match("[Cc]ookie") then
+      table.insert(cookie, line)
+    end
+  end
+
+  if #hsts_header > 0 then
+    table.insert(output_info, "HSTS is configured.")
+    table.insert(output_info, "Header: " .. table.concat(hsts_header, " "))
+  elseif port.number == 443 then
```

Comment:
An HTTPS server may be running on another port, triggering this script via Service Detection. It would be nicer if this line checked if this port was registered as using SSL by using `shortport.ssl`.